source/module_hamilt_general/module_xc/xc_functional_libxc_wrapper_tauxc.cpp

```diff
@@ -35,61 +38,50 @@ void XC_Functional::tau_xc(const double &rho, const double &grho, const double &
         v1xc += v1;
         v3xc += v3;
     }
-    finish_func(funcs);
+    XC_Functional_Libxc::finish_func(funcs);
 
 	return;
 }
 
 
-void XC_Functional::tau_xc_spin(double rhoup, double rhodw, 
+void XC_Functional_Libxc::tau_xc_spin(
+        const std::vector<int> &func_id,
+        double rhoup, double rhodw,
         ModuleBase::Vector3<double> gdr1, ModuleBase::Vector3<double> gdr2,
         double tauup, double taudw,
         double &sxc, double &v1xcup, double &v1xcdw, double &v2xcup, double &v2xcdw, double &v2xcud,
         double &v3xcup, double &v3xcdw)
 {
-
-	std::vector<xc_func_type> funcs = init_func(XC_POLARIZED);
-
-    double *rho, *grho, *tau, *v1xc, *v2xc, *v3xc, *sgn, s;
     sxc = v1xcup = v1xcdw = 0.0;
     v2xcup = v2xcdw = v2xcud = 0.0;
     v3xcup = v3xcdw = 0.0;
-    rho = new double[2];
-    grho= new double[3];
-    tau = new double[2];
-    v1xc= new double[2];
-    v2xc= new double[3];
-    v3xc= new double[2];
-    sgn = new double[2];
-    
-    rho[0] = rhoup;
-    rho[1] = rhodw;
-    grho[0] = gdr1.norm2();
-    grho[1] = gdr1 * gdr2;
-    grho[2] = gdr2.norm2();
-    tau[0] = tauup;
-    tau[1] = taudw;
 
-    double lapl[2] = {0.0,0.0};
-    double vlapl[2];
+    const std::vector<double> rho = {rhoup, rhodw};
+    const std::vector<double> grho = {gdr1.norm2(), gdr1 * gdr2, gdr2.norm2()};
+    const std::vector<double> tau = {tauup, taudw};
 
-    const double rho_threshold = 1E-6;
-    const double grho_threshold = 1E-10;
+	std::vector<xc_func_type> funcs = XC_Functional_Libxc::init_func(func_id, XC_POLARIZED);
 
     for(xc_func_type &func : funcs)
     {
         if( func.info->family == XC_FAMILY_MGGA || func.info->family == XC_FAMILY_HYB_MGGA)
         {
-            sgn[0] = sgn[1] = 1.0;
-            // call Libxc function: xc_mgga_exc_vxc
-            xc_mgga_exc_vxc( &func, 1, rho, grho, lapl, tau, &s, v1xc, v2xc, vlapl, v3xc);
+            constexpr double rho_threshold = 1E-6;
+            constexpr double grho_threshold = 1E-10;
+            std::vector<double> sgn = {1.0, 1.0};
             if(func.info->kind==XC_CORRELATION)
             {
                 if ( rho[0]<rho_threshold || sqrt(std::abs(grho[0]))<grho_threshold )
                     sgn[0] = 0.0;
                 if ( rho[1]<rho_threshold || sqrt(std::abs(grho[2]))<grho_threshold )
                     sgn[1] = 0.0;
             }
+
+            double s = 0.0;
+            std::vector<double> v1xc(2), v2xc(3), v3xc(2), lapl(2), vlapl(2);
```

Comment:
I suggest not use std::vector or new/delete for these variables which should be allocated on stack memory!

Reply:
updated